modules/core/src/main/scala/org/tessellation/infrastructure/dag/DAGService.scala

```diff
@@ -15,23 +19,44 @@ object DAGService {
   def make[F[_]: Async](globalSnapshotStorage: GlobalSnapshotStorage[F]): DAGService[F] =
     new DAGService[F] {
 
-      def getBalance(address: Address): F[Balance] =
-        globalSnapshotStorage.head.map {
-          case Some(snapshot) => snapshot.value.info.balances.get(address).getOrElse(Balance.empty)
-          case None           => Balance.empty
+      def getBalance(address: Address): F[Option[(Balance, SnapshotOrdinal)]] =
+        globalSnapshotStorage.head.flatMap {
+          case Some(snapshot) => getBalance(snapshot.value.ordinal, address)
+          case _              => none.pure[F]
         }
 
-      def getTotalSupply: F[BigInt] = {
+      def getBalance(ordinal: SnapshotOrdinal, address: Address): F[Option[(Balance, SnapshotOrdinal)]] =
+        globalSnapshotStorage.get(ordinal).map {
+          case Some(snapshot) =>
+            val balance = snapshot.value.info.balances.get(address).getOrElse(Balance.empty)
+            val ordinal = snapshot.value.ordinal
+
+            (balance, ordinal).some
+          case _ => none
+        }
+
+      def getTotalSupply: F[Option[(BigInt, SnapshotOrdinal)]] =
+        globalSnapshotStorage.head.flatMap {
+          case Some(snapshot) => getTotalSupply(snapshot.value.ordinal)
+          case _              => none.pure[F]
+        }
```

Comment:
Same here. With monad transformer we avoid explicit fallback from `None` to `F[None]`

```scala
OptionT(globalSnapshotStorage.head)
   .flatMap(latest => OptionT(getTotalSupply(latest.value.ordinal)))
   .value
```